File: Tree/100.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

struct TreeNode* newNode(int val) {
    struct TreeNode* node = malloc(sizeof(struct TreeNode));
    node->val = val;
    node->left = node->right = NULL;
    return node;
}

/* TODO: 你要自己實作這個函式 */
bool isSameTree(struct TreeNode* p, struct TreeNode* q){
    bool ans = false;
    if(!p && !q){
        return true;
    }
    if(!p || !q){
        return false;
    }

    if(p->val != q->val){
        return false;
    }

    if(p && q){
        bool left = isSameTree(p->left, q->left);
        bool right = isSameTree(p->right, q->right);
        ans = left && right;
    }
    return ans;

};

void freeTree(struct TreeNode* root) {
    if (!root) return;
    freeTree(root->left);
    freeTree(root->right);
    free(root);
}

int main(void) {
    /* 測試1: 相同的樹 */
    struct TreeNode* p1 = newNode(1);
    p1->left = newNode(2);
    p1->right = newNode(3);

    struct TreeNode* q1 = newNode(1);
    q1->left = newNode(2);
    q1->right = newNode(3);

    printf("Test1 => %s (true)\n", isSameTree(p1, q1) ? "true" : "false");

    freeTree(p1);
    freeTree(q1);

    /* 測試2: 不同的樹 */
    struct TreeNode* p2 = newNode(1);
    p2->left = newNode(2);

    struct TreeNode* q2 = newNode(1);
    q2->right = newNode(2);

    printf("Test2 => %s (false)\n", isSameTree(p2, q2) ? "true" : "false");

    freeTree(p2);
    freeTree(q2);

    return 0;
}

File: Tree/100.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSameTree(self, p, q):
        """
        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """
        # TODO: 你寫
        if not p and not q:
            return True
        if not p or not q:
            return False
        if p.val == q.val:
            return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
        else:
            return False
        



# -------------------------
# 測試用樹
# -------------------------

def build_same_tree():
    #     1
    #    / \
    #   2   3
    p = TreeNode(1)
    p.left = TreeNode(2)
    p.right = TreeNode(3)

    q = TreeNode(1)
    q.left = TreeNode(2)
    q.right = TreeNode(3)

    return p, q


def build_diff_tree():
    #     1        1
    #    /          \
    #   2            2
    p = TreeNode(1)
    p.left = TreeNode(2)

    q = TreeNode(1)
    q.right = TreeNode(2)

    return p, q


if __name__ == "__main__":
    sol = Solution()
    
    p, q = build_same_tree()
    print(sol.isSameTree(p, q))  # True

    p, q = build_diff_tree()
    print(sol.isSameTree(p, q))  # False